### PR TITLE
[WIP] Jump to node

### PR DESCRIPTION
**Pull request by ohuang12**
Adds the options in the menu to:
- Jump to a node, based on its id.
- Jump to a certain rank.
- Jump to nodes which are attached to annotations with a certain gene name.

Additionally adds the setup for a slider (visibility turned off).

### CODE REVIEW

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WindowController.java [hunk 1]
```diff
@@ -27,32 +32,40 @@
...
 	}
 	
-	private void constructView() {
+	private void constructView(Pane strainView, Pane phyloView) {
 		SplitPane splitPane = new SplitPane(strainView, phyloView);
 		splitPane.setOrientation(Orientation.VERTICAL);
 		root.setCenter(splitPane);
 	}
 
+	private Pane createStrainView(ColorServer colorServer, Graph graph) {
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WindowController.java]: You don't necessarily need to do this. The minimap I'm working on ATM gets created and added inside `StrainView` itself, which still is just a regular `Pane`. I'm thinking that you can also create `StrainControl` directly from `StrainView` itself. You can take a look at [the minimap](https://github.com/Unia/dnainator/blob/minimap/dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/StrainView.java#L27) to see how it's done.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WindowController.java [hunk 2]
```diff
@@ -71,21 +84,57 @@ private void exitAction(ActionEvent e) {
 
 	@SuppressWarnings("unused") @FXML
 	private void zoomInAction(ActionEvent e) {
-		strainView.zoomIn();
+		if (strainView != null) {
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WindowController.java]: Why are these needed all of a sudden?

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/AbstractView.java [hunk 3]
```diff
@@ -41,14 +41,22 @@ public AbstractView() {
...
+	 */
+	public void setScale(double scale) {
+		this.scale = new Affine(new Scale(scale, scale));
+	}
+	
+	/**
 	 * @return The scale used to scale the contents of the {@link AbstractView}.
 	 */
 	public Affine getScale() {
-		return new Affine(new Scale(SCALE, SCALE));
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/AbstractView.java]: Why is this change, and the addition of `setScale` needed?

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/StrainView.java [hunk 4]
```diff
@@ -52,4 +59,47 @@ public void resetTranslate() {
...
 		updateStrain();
 	}
+	
+	/**
+	 * Sets the panning of the {@link StrainView}.
+	 * @param x the amount to pan on the x axis.
+	 * @param y the amount to pan on the y axis.
+	 */
+	public void setPan(double x, double y) {
+		resetZoom();
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/StrainView.java]: You can restore the current zoomlevel by saving it, see [here](https://github.com/Unia/dnainator/blob/minimap/dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/StrainView.java#L54) in my minimap branch.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/widgets/StrainControl.java [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+		jumpTo.setPromptText(name);
+	}
+	
+	private void setupImage() {
+		Image image = new Image(MAGGLASS);
+		magGlass = new ImageView();
+		magGlass.setImage(image);
+		magGlass.setSmooth(true);
+		magGlass.setCache(true);
+		// CHECKSTYLE.OFF: MagicNumber
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/widgets/StrainControl.java]: Please replace these with the usual stuff.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/widgets/StrainControl.java [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+		});
+	}
+	
+	/**
+	 * Changes the prompt text of the {@link TextField} and marks it in red.
+	 */
+	private void promptInvalid() {
+		jumpTo.clear();
+		jumpTo.setPromptText(ERROR);
+		jumpTo.setStyle("-fx-prompt-text-fill: red;");
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/widgets/StrainControl.java]: Might want to move this to css

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/widgets/StrainControl.java [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+				promptInvalid();
+			}
+		});
+	}
+	
+	/**
+	 * Setup the jump to gene {@link TextField}.
+	 * @return the text field of the jump to gene.
+	 */
+	private void createJumpToGene() {
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/widgets/StrainControl.java]: I think your methods are slightly too long...

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WindowController.java (hunk 2 above)
Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WindowController.java]: You get NPs when you wish to zoom in before the strainview is created (in the welcome screen).

Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WindowController.java]: Then isn't it better to disable those options instead?

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/StrainView.java (hunk 4 above)
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/StrainView.java]: It's not exactly restoring anymore, but the `setPan` from my branch is the one we want. It does not require the zoom to be reset :)

**Comment by Unia** [comment]
I think the controls should all be visible by default, and I think that we can drop the whole `Navigate` submenu. Or we can keep it, in which case the menu options focus the relevant text entry. I also think we can drop the magnifying glass: it doesn't add any value and it's out of style with the rest.

Also, can you reduce the spacing between the text entry (entries) and the stepper?

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/AbstractView.java (hunk 3 above)
Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/AbstractView.java]: Not necessarily needed. I thought this would be more consistent with other getters, as it would return the actual currently used scale.

Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/AbstractView.java]: Please do that in a separate PR then. It adds noise to this one.

File: dnainator-core/src/main/java/nl/tudelft/dnainator/graph/impl/Neo4jGraph.java [hunk 8]
```diff
@@ -124,6 +124,9 @@ public AnnotationCollection getAnnotations() {
 	 * @return a {@link SequenceNode} with the information of the given {@link Node}.
 	 */
 	public EnrichedSequenceNode createSequenceNode(Node node) {
+		if (node == null) {
```
Comment (gfokkema) [inline dnainator-core/src/main/java/nl/tudelft/dnainator/graph/impl/Neo4jGraph.java]: This check does not belong here, this method should never be passed a null value...

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WindowController.java [hunk 9]
```diff
@@ -34,7 +40,7 @@ private void initialize() {
 		fileOpenController.dbPathProperty().set(welcomeController.getListedPaths());
 		fileOpenController.graphProperty().addListener((obj, oldV, newV) -> {
 			strainView = new StrainView(colorServer, newV);
-			constructView();
+			constructView(strainView, phyloView);
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WindowController.java]: Isn't `phyloView` still `null` here?

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/AbstractView.java [hunk 10]
```diff
@@ -161,8 +173,7 @@ public void zoomOut() {
 	 * Resets the zoom level to the default value.
 	 */
 	public void resetZoom() {
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/AbstractView.java]: This is unused now, isn't it?

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/StrainView.java [hunk 11]
```diff
@@ -52,4 +68,49 @@ public void resetTranslate() {
 		super.resetTranslate();
 		updateStrain();
 	}
+	
+	/**
+	 * Sets the panning of the {@link StrainView}.
+	 * @param x the amount to pan on the x axis.
+	 * @param y the amount to pan on the y axis.
+	 */
+	public void setPan(double x, double y) {
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/StrainView.java]: Perhaps this should be made private?

Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/StrainView.java]: Yes, better if it is. Will update.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WindowController.java (hunk 9 above)
Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/controllers/WindowController.java]: No it isn't, however it is no longer needed to pass these as parameters.

**Comment by gfokkema** [comment]
The slider does really weird stuff, either remove the zoom slider or fix it before we can merge this PR.

**Comment by Unia** [comment]
I vote for removing the slider and focusing on more pressing tasks, such as consistent CSS styling.

**Comment by gfokkema** [comment]
Agreed.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/StrainView.java [hunk 12]
```diff
@@ -35,6 +43,14 @@ public void pan(Point2D delta) {
 		updateStrain();
 	}
 
+	/**
+	 * Zoom based on center for stepper.
+	 * @param delta the amount to zoom in.
+	 */
+	public void zoom(double delta) {
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/StrainView.java]: Can be removed now that the stepper is gone.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/StrainView.java [hunk 13]
```diff
@@ -52,4 +68,49 @@ public void resetTranslate() {
...
+	 */
+	public void gotoRank(int rank) {
+		setPan(-rank * strain.getRankWidth(), 0);
+	}
+	
+	/**
+	 * Center view on a node given the id.
+	 * @param id the id of the node.
+	 */
+	public void centerNode(String id) {
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/StrainView.java]: `centerNodeVertically`?

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/widgets/StrainControl.java [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+	public StrainControl(StrainView strainView) {
+		this.strainView = strainView;
+		setPadding(new Insets(PADDING));
+		setSpacing(PADDING);
+		setMaxWidth(WIDTH);
+		setPickOnBounds(false);
+		jumpTo = new TextField();
+		curGraph = strainView.getStrain().getGraph();
+		
+		// Wait for strain view to be created.
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/widgets/StrainControl.java]: Can't this run sooner than that the strain view is created?

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/widgets/StrainControl.java [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+	 * Setup the jump to node {@link TextField}.
+	 * @return the text field of the jump to node.
+	 */
+	private void createJumpToNode() {
+		setupTextField(NODE);
+		jumpTo.setOnAction(e -> jumpToNode());
+	}
+	
+	private void jumpToNode() {
+		String inputText = jumpTo.getCharacters().toString();
```
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/widgets/StrainControl.java]: @gfokkema had another solution for this, hadn't he?

**Comment by Unia** [comment]
It's working nice now! :) Just some small comments still...

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/StrainView.java (hunk 13 above)
Comment (ohuang12) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/views/StrainView.java]: Yes, seems like a more appropriate name.

File: dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/widgets/StrainControl.java (hunk 15 above)
Comment (Unia) [inline dnainator-javafx/src/main/java/nl/tudelft/dnainator/javafx/widgets/StrainControl.java]:
```
int rank = curGraph.getNode(inputText).getProperty("RANK");
resetPromptText(NODE);
strainView.gotoRank(rank;
```

Is that it, @gfokkema ?

**Comment by Unia** [comment]
Superseded by #159. Closing.
